pkg/tfshim/sdk-v2/provider2.go

```diff
@@ -322,31 +323,70 @@ func (p *planResourceChangeImpl) upgradeState(
 ) (shim.InstanceState, error) {
 	res := p.tf.ResourcesMap[t]
 	state := p.unpackInstanceState(t, s)
-	instanceState, err := res.ShimInstanceStateFromValue(state.stateValue)
+
+	ty := res.CoreConfigSchema().ImpliedType()
+
+	jsonIsh, err := schema.StateValueToJSONMap(state.stateValue, ty)
```

Comment:
Calling `schema.StateValueToJSONMap` looses information. The implementation of `schema.StateValueToJSONMap` is:
```go
// StateValueToJSONMap converts a cty.Value to generic JSON map via the cty JSON
// encoding.
func StateValueToJSONMap(val cty.Value, ty cty.Type) (map[string]interface{}, error) {
	js, err := ctyjson.Marshal(val, ty)
	if err != nil {
		return nil, err
	}

	var m map[string]interface{}
	if err := json.Unmarshal(js, &m); err != nil {
		return nil, err
	}

	return m, nil
}
```

Reply:
Which information is lost here?

Reply:
Appreciate unpacking the concern a bit so we can address!

This is extracted from #1971  which we started taking as far as downstream tests.

I noticed in your previous rewrite @iwahbe you prefer passing the information via flatmap 

https://github.com/pulumi/pulumi-terraform-bridge/pull/1735/files#diff-1c55ab32bd51c4f37c224239b498efca8a5b19c33cbb244231a5d6f02d301affR44

IN here we don't have easy access to flatmap but we're working with cty.Value:

```
type v2InstanceState2 struct {
	resourceType string
	stateValue   cty.Value
	// Also known as private state.
	meta map[string]interface{}
}
```

If there's a way to go from cty.Value to flatmap that's better than going from cty.Value to json-ish we could use it here.

Reply:
We don't need to use flatmap. I used it previously because that was what we were doing before. We should use JSON for inputs.

Reply:
I'm thinking state.stateValue.Type() can backfire as well as it can't get schema Set functions from schema etc.. Eh, no good options here until https://github.com/pulumi/pulumi-terraform-bridge/issues/1667 lands. Should at least comment and link that.